### PR TITLE
Emit metadata files earlier

### PR DESCRIPTION
**Pull request by nnethercote**
This will make cargo pipelining much more effective.

### CODE REVIEW

**Comment by nnethercote** [comment]
r? @nnethercote 

...because this code isn't ready for review yet.

**Comment by nnethercote** [comment]
r? @alexcrichton 

I have moved metadata encoding and writing to the start of `start_codegen`, which is right at the start of code generation.

You hinted that it might be possible to move it even earlier, before analysis. I haven't done that because it's fiddly (even moving it as far as I did was fiddly) and I didn't want to do it without having a clearer idea of whether it would work. Also, writing metadata before doing analysis would penalize the speed of compiler invocations that end in compile errors (which are common), so it's not a clear win.

Anyway, take a look, see what you think about what I've done so far.

**Comment by alexcrichton** [comment]
@bors: r+

Looks great to me!

I think it's actually a good point that we should wait to commit metadata until the crate has passed most static analysis anyway, in which case writing it just as codegen starts I think is at the very least a great start for now.

**Comment by mark-i-m** [comment]
Does it make sense to move it earlier only for non-check builds or only for release builds?

**Comment by nnethercote** [comment]
> Does it make sense to move it earlier only for non-check builds or only for release builds?

If people religiously used `cargo check` to get past all their compile errors and only then used `cargo build`, I would say yes. But I doubt that's true. Let's see how this choice pans out, we can revisit later if it seems worthwhile.

**Comment by mark-i-m** [comment]
Fair enough, though it seems reasonable to only promise really fast builds if one uses check...

File: src/librustc_codegen_llvm/lib.rs [hunk 1]
```diff
@@ -110,12 +110,13 @@ impl ExtraBackendMethods for LlvmCodegenBackend {
         ModuleLlvm::new_metadata(tcx, mod_name)
     }
 
-    fn write_metadata<'b, 'gcx>(
+    fn write_compressed_metadata<'b, 'gcx>(
```
Comment (eddyb) [inline src/librustc_codegen_llvm/lib.rs]: I think this name is misleading, it should probably be `{inject,embed,include,etc.}_compressed_metadata`.

**Pull request merged**